jlh/WeatherApp/app/src/main/java/com/e_conomic/weatherapp/ui/ForecastListAdapter.kt

```diff
@@ -0,0 +1,43 @@
+package com.e_conomic.weatherapp.ui
+
+import android.support.v7.widget.RecyclerView
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import com.e_conomic.weatherapp.R
+import com.e_conomic.weatherapp.domain.model.Forecast
+import com.e_conomic.weatherapp.domain.model.ForecastList
+import com.squareup.picasso.Picasso
+import ctx
+import kotlinx.android.synthetic.main.item_forecast.view.*
+
+class ForecastListAdapter(val weekForecast: ForecastList, val itemClick: (Forecast) -> Unit) :
+        RecyclerView.Adapter<ForecastListAdapter.ViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(parent.ctx).inflate(R.layout.item_forecast, parent, false)
+        return ViewHolder(view, itemClick)
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bindForecast(weekForecast[position])
+    }
+
+    override fun getItemCount() = weekForecast.size
+
+    class ViewHolder(view: View, val itemClick: (Forecast) -> Unit) :
+            RecyclerView.ViewHolder(view) {
+
+        fun bindForecast(forecast: Forecast) {
+            with(forecast) {
+                Picasso.with(itemView.ctx).load(iconUrl).into(itemView.icon)
```

Comment:
Yes it would do that, but with a caveat. The note in the documentation hints that `Picasso` addresses the issue though.

Imagine the following sequence of events:
Bind forecast_1
Async load image_1
View displays forecast_1 date etc (no image yet)
-- ListView is scrolled and view recycled ---
Bind forecast_2
Async load image_2
View displays forecast_2 date etc (no images yet)
Async image_2 is loaded and displayed
Async image_1 is loaded and displayed

=> forecast_2 is diplayed with image_1

What Picasse must be doing under the hood is storing the `iconUrl` with the imageView, and an image is asynchronously downloaded, check that the downloaded image-url, is still that of the `iconUrl` set on the imageView. In the steps above, this check would mean that image_1 would not be set in the final step. Also, `Picassa` probably does (at least it should) queues and cancels requests smartly.